### PR TITLE
Disable rendering highly transparent elements

### PR DESCRIPTION
**Pull request by sarbes**
## Description
This patch disables rendering of highly transparent GUI elements. The threshold is at 99% transparency.

## Motivation and context
In some scenes in Estuary (and on others, I presume), textures and text might be faded away while still being rendered. This has a negative GPU performance impact.

## How has this been tested?
One obvious scene where this patch shines is the info dialog of Estuary. As you can see in the screenshots, the amount of overdraw is hugely reduced. The amount of draw calls is halved (160 vs. 80). According to Renderdoc, the rendering time went from 1.74ms to 1.17ms (-33%) on my RX570 at 4k.

Another benefit I could observe is on the Estuary home screen. In between the category transition animation (where the old set of element fades out while the new one fades in), zero to one elements get rendered at all times. Before, both elements would get rendered.

I couldn't find any rendering issues.

## What is the effect on users?
It shouldn't have any impact except improved GPU performance.

## Screenshots (if appropriate):
The Estuary info dialog:
infoscreen
Overdraw before the patch (notice the home screen in the background)
before_overlay
Overdraw after the patch:
after_overlay

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by sarbes** [comment]
On a side note, dirty regions don't reflect this change. Such a change is not totally trivial and might give only a tiny payoff.

**Comment by garbear** [comment]
This is a creative optimization. 99% seems like a good value. Too bad DR is more complicated to optimize.

**Comment by sarbes** [comment]
I'm not entirely sure how the animation system works, but the savings with dirty regions should be not as large as this optimization. The case where savings could be made would be when the element is still moving while being faded out. Once the animation stops, the area is not dirty anymore. And even then, it is usually just the background which has to be drawn.

And let's face it - there are more glaring issues in our DR system.

**Comment by sarbes** [comment]
On further thought, we might be able to do the check here: https://github.com/xbmc/xbmc/blob/99a4d44eb55763234bc9a4ffce471e1a039ebfed/xbmc/guilib/GUIControl.cpp#L403

This seems to include the dirty regions. But I'm not sure if this breaks something. Seems okay on first glance. Opinions?

**Comment by AlwinEsch** [comment]
Looks like a good idea for me :+1:. Maybe test also it by the screensaver and visualization add-ons within Kodi's window changes :thinking: think does not matter to it, but to make sure.

**Comment by garbear** [comment]
Transform is both a rendering property and a visibility property, because transforming to an alpha of <1% means invisible. So I would move the check inside `IsVisible()`. That's just my logic, of course we'll have to see the results.

**Comment by sarbes** [comment]
Huh, pushed the branch with no changes by accident. It should be fine now, I've moved the check.

File: xbmc/guilib/GUIControl.cpp [hunk 1]
```diff
@@ -403,6 +403,8 @@ bool CGUIControl::CanFocus() const
 bool CGUIControl::IsVisible() const
 {
   if (m_forceHidden) return false;
+  if (m_transform.alpha <= 0.01f)
```
Comment (enen92) [inline xbmc/guilib/GUIControl.cpp:406]: Maybe combine with the condition above instead of adding another if/branch?

Comment (sarbes) [reply xbmc/guilib/GUIControl.cpp:406]: If you want. I've did it in a similar style as above.

**Review by garbear — APPROVED** [review]
Look good to me now

The only issue I could think is if a skin purposefully set a control to 0.01f to make it "see-through" but not fail an "is-visible" condition. But I guess that's the skin's problem if it does this confusing thing now.

**Comment by sarbes** [comment]
What would be the use case of such a behaviour? Sounds like a hack to me.

Overall, I don't think such a behaviour is not needed if skins would set the visibility condition on fade-out. But I have not seen any skin doing this, as it has no influence on the output image.

**Review by enen92 — APPROVED** [review]

File: xbmc/guilib/GUIControl.cpp (hunk 1 above)
Comment (sarbes) [reply xbmc/guilib/GUIControl.cpp]: Done.

**Comment by sarbes** [comment]
Thanks guys.

**Comment by enen92** [comment]
I think this PR should be reverted and replaced by an alternative that doesn't touch common logical functions such as `IsVisible()` used pretty much everywhere on the GUI engine. If possible try to find some alternative that only requires usage in Render/DoRender

**Pull request merged**